src/shadowcasting.cpp

```diff
@@ -0,0 +1,707 @@
+                        // Pass true to the span constructor to set skip_first_row to true
+                        // This prevents the same row we are currently checking being check by the
+                        // new D span
+                        spans.emplace( std::next( this_span ),
+                                       leading_edge_major, this_span->end_major,
+                                       this_span->start_minor, this_span->end_minor,
+                                       this_span->cumulative_value, true );
+                    }
+
+                    // Truncate this_span to the current block.
+                    this_span->start_major = std::max( this_span->start_major, trailing_edge_major );
+                    this_span->end_major = std::min( this_span->end_major, leading_edge_major );
+
+                    // The new span starts at the leading edge of the previous square if it is opaque,
+                    // and at the trailing edge of the current square if it is transparent.
+                    this_span->start_minor = new_start_minor;
+                    this_span->cumulative_value = next_cumulative_transparency;
+
+                    new_start_minor = leading_edge_minor;
+                    current_transparency = new_transparency;
+                }
+
+                // If we end the row with an opaque tile, set the span to start at the next row
+                // since we don't need to process the current one any more.
+                if( !check( current_transparency, last_intensity ) ) {
+                    this_span->start_major = leading_edge_major;
+                }
+            }
+
+            if( !started_block ) {
+                // If we didn't scan at least 1 z-level, don't iterate further
+                // Otherwise we may "phase" through tiles without checking them or waste time
+                // checking spans that are out of bounds.
+                this_span = spans.erase( this_span );
+            } else if( !check( current_transparency, last_intensity ) ) {
+                // If we reach the end of the span with terrain being opaque, we don't iterate further.
+                // This means that any encountered transparent tiles from the current span have been
+                // split off into new spans
+                this_span = spans.erase( this_span );
+            } else {
+                // Cumulative average of the values encountered.
+                this_span->cumulative_value = accumulate( this_span->cumulative_value,
+                                              current_transparency, distance );
+                ++this_span;
+            }
+        }
+    }
+}
+
+template<int x_transform, int y_transform, int z_transform, typename T,
+         T( *calc )( const T &, const T &, const int & ),
+         bool( *check )( const T &, const T & ),
+         T( *accumulate )( const T &, const T &, const int & )>
+void cast_vertical_zlight_segment(
```

Comment:
Splitting `cast_zlight_segment` into vertical and horizontal versions isn't ideal and duplicates a lot of complex code.  However, the alternative is some quite nasty iteration logic to handle the full range of possible coordinate transformations instead of the subset currently handled.  Splitting it like this makes this already complex code a good bit more readable without the extra conditonals.

Would definitely appreciate feedback on whether this choice makes sense to others though.

Reply:
It's certainly possible to merge.  You'd want to abstract out the coordinate system more into some type template argument that would replace the various `_transform` numbers.  Each version of that type would have methods to do the various transformations needed (between depth/major/minor coordinates and x/y/z coordinates).

I think it is worth doing and would probably help clarity and maintainability, but it doesn't have to be on this first PR.